### DISCUSSION TITLE
Calculated Fields

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3079] [Feature Requests]
An _alias_ field that renders a calculated value, live but with some level of caching for performance gains.

**Comment by rijkvanzanten** [comment]
How does this tie in with directus/directus#2989? Would we be able to extend directus/directus#2989 to support "custom formulas"?

**Reply by rijkvanzanten** [reply]
I feel like these are two solutions to a similar use case. Though aggregate functions run on the whole "column", where this would be row to row?

**Reply by benhaynes** [reply]
Hmm, perhaps. I do see how these both exist as formulas, especially when looking at them as a spreadsheet style syntax. However the aggregate functions (when vanilla) can be run directly against the database, and very much seem like **column** metadata. What I'm thinking here is more **record** based (eg: a new field that runs calculations from other field values).

**Reply by benhaynes** [reply]
In short, the aggregate functions seem like a good thing for the sidebar... but calculated fields would be in the actual data set.

**Reply by theonlynexus** [reply]
I haven't yet dove into the code but, since there is support for read-only fields,  wouldn't calculated fields set directly in the table definition be just a special case of read-only field where the read-only attribute cannot be removed and that doesn't get updated?
If that's what an alias field is, then I beg your pardon in advance.

**Reply by rijkvanzanten** [reply]
@theonlynexus Yes and no, that would work great for calculations on the _row_ level (eg `fieldA + fieldB as column fieldC`), but won't work for things like "average value of fieldA across rows 1 thru 100") as those are on the "request level" and not the "row level"

**Comment by arladmin** [comment]
Hi there. This is a much sought after feature for most of my use-cases; the ability to have equations (numeric/date/text) [simple/javascript logic] as fields.
Is there already an ETA for it?

**Reply by benhaynes** [reply]
[**Unfortunately, we can not provide estimates on features or fixes**](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍 . Or contribute a pull-request... after all, this is OSS!

**Reply by somu22081980** [reply]
@benhaynes , I am using directus 9, the REST API of directus9is it supports Aggregate functions like SUM, AVERAGE etc?

**Reply by benhaynes** [reply]
Yes, but on a branch... it hasn't been released yet.

**Reply by somu22081980** [reply]
Thank you for your quick response, when it will be released? any dates?

**Reply by benhaynes** [reply]
No specific times, but hopefully in the next few weeks!

**Reply by somu22081980** [reply]
Thank you.

**Reply by somu22081980** [reply]
Dear Benhaynes, 

I hope following features are going to release or release in this week, can you please let us know how can we upgrade existing directus instance with these features?

![Directus New Features Image](https://user-images.githubusercontent.com/16229005/133724151-d96fd686-7768-4885-a49c-feffbd8d6ebf.JPG)

Regards,
Somaraju

**Reply by rijkvanzanten** [reply]
Hi @somu22081980. It depends on how you deploy Directus specifically, but you should be able to simply update to the latest version on NPM and run the database migrations afterwards: https://docs.directus.io/guides/projects/#upgrading-updating-a-project

**Comment by benhaynes** [comment]
@somu22081980 — are you self hosting or on our managed cloud?

**Comment by somu22081980** [comment]
Self hosting

**Reply by benhaynes** [reply]
You can follow our documentation on upgrading your project, and you'll get any features on the latest release. Curious why you commented on this ticket, to be clear, this feature is not currently released or on the immediate docket.

**Comment by somu22081980** [comment]
Ok thank you.

**Comment by veryphatic** [comment]
Is there any update on this feature? Or a branch that needs a little love? Happy to help :)

**Reply by benhaynes** [reply]
[**Unfortunately, we can not provide estimates on features or fixes**](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍 . Or contribute a pull-request... after all, this is OSS!

-----

_This is an automated response._

**Reply by vahidreza** [reply]
@benhaynes i assume by setting a webhook while logging the first request time to prevent a loop of calls ( or the hash of the record to detect a change and dropping recursive calls) and using js API we can calculate the field inside of a nodejs service and update the calculated column accordingly. Is it viable you think?

**Reply by azrikahar** [reply]
@vahidreza that is indeed towards the right general direction, but you can instead use [Custom Hooks](https://docs.directus.io/extensions/hooks/) to avoid needing the webhook (different from hooks) and not need the JS API in question to do sort of round trip requests 👍

Here's a sample another member of our community the other day which should be helpful to give a better idea on it: https://github.com/directus/directus/discussions/11422#discussioncomment-2109615

Basically they are doing _pseudo_ "calculated" field for slug, where it's a combination of parent & current input name. However this is still not real time or live of course.

Sort of a tangent to this, but there's also a feature in the works that may or may not help to make the above use case easier to achieve! 😊

**Reply by vahidreza** [reply]
@azrikahar tnx for the clarification, this is a way better approach.

**Comment by Voldemorten** [comment]
A temporary workaround until above is implemented, could be to create a custom hook. For instance, I have a "calculation field" / alias where I need number of days since creation. I have created a hook, that looks like this. 
```
var index = ({ filter, action }, { services, exceptions }) => {
    action('general_item.items.read', ({ payload }) => {
        for (let i = 0; i < payload.length; i++) {
            const item = payload[i];
            if (item.days_listed !== undefined) {
                const diffInMs = new Date() - new Date(item.date_created);
                item.days_listed = Math.floor(diffInMs / (1000 * 60 * 60 * 24));
            }
        }
    });
};
```
days_listed is a read-only input field, which will exist in the database, but it will always be empty in the actual database, as it is calculated only when the item is read.
Of course this degrades performance a bit, but if you do the check `if (item.days_listed !== undefined)`, you will only calculate the field when it's actually necessary. Performance could further be improved by caching. 

Does anyone see a problem with this approach?

**Reply by Voldemorten** [reply]
There is a problem with the above approach if you are using nested items and api calls.
In my scenario, I have a `gold_item` which have a relation to a `general_item`. If I use the following api call
`http://localhost:8055/items/gold_item?fields=*.*&access_token=xxx`
I get the general_item back, but days_listed is not calculated, because the `action('general_item.items.read', ...)` is never run. It seems like the read event is not emitted, when we do a nested read.

![image](https://user-images.githubusercontent.com/8776117/153875237-1978f053-97e9-46e5-a514-94e56b22f5b1.png)

Can anyone elaborate on this?

**Reply by rijkvanzanten** [reply]
> is never run. It seems like the read event is not emitted, when we do a nested read.

This is correct, as the nested collection item isn't read directly. All of the nested relational data is fetched in the least amount of database queries possible[^1] rather than doing a new read operation against the items service per nested level. This also means that the filter is never fired, as that read operation is never performance against the ItemsService.

[^1]: Except for when less queries is worse for performance :) https://github.com/directus/directus/issues/11873

**Reply by Voldemorten** [reply]
That makes sense. Thank you.

**Reply by Voldemorten** [reply]
@rijkvanzanten do you have any high level ideas on how we could solve the issue with calculated fields. Right now I have a solution, where I have created an interface, where you can input custom JS code, and refer to the item within. This looks like this as we speak: 
![image](https://user-images.githubusercontent.com/8776117/156149626-f4b45a67-144a-40be-adef-cff47895c3fc.png)
This is coupled to a hook on the specific item type, that basically looks at all fields of type "calculation" and if the value for this is empty, it does the calculation instead. But again, this doesn't work for nested items. Another solution could be to have the same hook, but have two scenarios: 
1. The current value is empty: We do the calculation and update the database with the result.
2. The current value is not empty: a) we do the calculation, and if the results match, we do nothing. b) we do the calculation and if the results doesn't match, we write the new calculation to the item. 

What do you think about the above solution?
I don't particularly like the idea of custom hooks that you need to create for each collection, but I can't figure out to do this just in the interface, without tinkering with the core of directus too much. 

If you could point me in the right direction, maybe I could look into a possible way of solving this issue. (That also works with nested items).

**Reply by rijkvanzanten** [reply]
Truly living up to the gif on your profile with this one 😄 

I think the best case scenario here (besides actually implementing native calculated fields support) would be to have a single hook[^1] that allows you to alter the read data based on the AST that's constructed internally. That should allow you to recursively find the collections/fields you want to transform, which in turn allows you to more easily modify the output data.

[^1]: Right now, the existing `read` hook fires before the AST is constructed, so we might want to look into adding a dedicated "after database query" filter hook that contains the ran AST + the output data so you can modify it closer to the source 🤔

**Reply by Voldemorten** [reply]
When you say "a single hook", do you mean a hook for each collection that contains "calculated fields"? Because I can't see that I can set up a global read hook on all collections, as directus is now. 

And if someone were to implement native calculated fields, which course of action would you recommend? (Not saying, that I'm going to look into it, but it would be nice to have an idea on how it would/should be implemented)

**Reply by rijkvanzanten** [reply]
That's correct, it would be a single hook that would fire everytime a read query is made to the database, so you can modify it right before it's converted to SQL and fired to the database.

> And if someone were to implement native calculated fields, which course of action would you recommend? (Not saying, that I'm going to look into it, but it would be nice to have an idea on how it would/should be implemented)

For native calculated fields, I'd want to rely on database native calculated fields (through generation expressions). In order to configure those from Directus itself, we would have to come up with some wrapper language that would allow for all the popular database functions, without having to write the vendor specific SQL

**Reply by Voldemorten** [reply]
> That's correct, it would be a single hook that would fire everytime a read query is made to the database, so you can modify it right before it's converted to SQL and fired to the database.

How would you create the single hook as directus is now? Would you separate the function and then do this for every collection `(<collection>.)items.read`? 
i.e. 
```
export default async ({ action }, { services, exceptions, database }) => {
    action(
        'customers.items.update',
        sharedHook()
    );
    action(
        'orders.items.read',
        sharedHook()
    );
    ...
    action(
        'generic_collection.items.read',
        sharedHook()
    );
};
```

**Reply by rijkvanzanten** [reply]
The way Directus currently works is that the request that's made is transformed into a JSON representation of the structure + query you want to make (AST) which is then parsed by a function to run in sql. I think to do "proper", I'd like to see a new hook that sits right in between that ASL->SQL step, so you can really tweak the nitty gritty, and modify it based on the AST made 🤔 

(By the way, `items.update` is the shorthand for all, you're not required to scope the collection)

**Comment by somu22081980** [comment]
Team,

I am looking how can we configure hooks in Directus Cloud environment?

Regards,
Somaraju


On Tue, Mar 1, 2022, 21:40 Rijk van Zanten ***@***.***> wrote:

> Truly living up to the gif on your profile with this one 😄
>
> I think the best case scenario here (besides actually implementing native
> calculated fields support) would be to have a single hook1
> <#m_8064591524824211310_user-content-fn-1-b305bfd91162c71597c0156c6189105e>
> that allows you to alter the read data based on the AST that's constructed
> internally. That should allow you to recursively find the
> collections/fields you want to transform, which in turn allows you to more
> easily modify the output data.
> Footnotes
>
>    1.
>
>    Right now, the existing read hook fires before the AST is constructed,
>    so we might want to look into adding a dedicated "after database query"
>    filter hook that contains the ran AST + the output data so you can modify
>    it closer to the source 🤔 ↩
>    <#m_8064591524824211310_user-content-fnref-1-b305bfd91162c71597c0156c6189105e>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/directus/directus/discussions/3079#discussioncomment-2274520>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD32FDIEY4HCOK2Y7DID7PTU5Y6PXANCNFSM4U62G4DA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
IBOT CONFIDENTIAL: The contents of this e-mail message and any attachments 
are confidential and are intended solely for the addressee. The information 
may also be legally privileged. This transmission is sent in trust, for the 
sole purpose of delivery to the intended recipient. If you have received 
this transmission in error, any use, reproduction or dissemination of this 
transmission is strictly prohibited. If you are not the intended recipient, 
please immediately notify the sender by reply e-mail or phone and delete 
this message and its attachments, if any.

**Reply by rijkvanzanten** [reply]
Custom code extensions are available on the Enterprise tier. You can contact the cloud team through https://directus.io/contact for more information 🙂

**Comment by somu22081980** [comment]
Dear Team,

Thank  you for your quick response, I have another query, where I am
trying to whitelist IPs, which is not working, please check the following
image for your more information.

And please clarify the following, the IP whitelisting is only for Admin UI
or APIs also (if use the access token generated from the user where IP is
whitelisted).

[image: image.png]


Kindly confirm us the same.

Best Wishes,
*Somaraju Pitchika*
Head, Delivery

*iBot Control Systems India P.Ltd. *Hitech City, Hyderabad, India

P: +91 9949-582-404
E: ***@***.*** ***@***.***>
W: http://www.iamibot.com

*Create the Internet of Your Things with iBot!*


On Tue, Mar 1, 2022 at 10:49 PM Rijk van Zanten ***@***.***>
wrote:

> Custom code extensions are available on the Enterprise tier. You can
> contact the cloud team through https://directus.io/contact for more
> information 🙂
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/directus/directus/discussions/3079#discussioncomment-2275026>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD32FDJ52DGV4WWE5XXG6NLU5ZGSFANCNFSM4U62G4DA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
IBOT CONFIDENTIAL: The contents of this e-mail message and any attachments 
are confidential and are intended solely for the addressee. The information 
may also be legally privileged. This transmission is sent in trust, for the 
sole purpose of delivery to the intended recipient. If you have received 
this transmission in error, any use, reproduction or dissemination of this 
transmission is strictly prohibited. If you are not the intended recipient, 
please immediately notify the sender by reply e-mail or phone and delete 
this message and its attachments, if any.

**Reply by rijkvanzanten** [reply]
Heya! This sounds unrelated to the current discussion. Mind opening a new one? Thanks!

**Comment by nordcart** [comment]
I have one field with the data(string) "1/2022" which means january 2022. 
The deadline for this is always 2 months later and the 12th day (march the 12th).

Would it be possible to calculate this from a string? And then for example show a red warning symbol or something when it is past the due date?

**Reply by benhaynes** [reply]
Seems pretty possible with a custom display.

**Reply by nordcart** [reply]
Is this already possible?

**Comment by stx-chris** [comment]
I would like to revive this topic 🤠. Any workaround involving hooks has its downsides and results in a lot of boilerplate, so I had rather have computed fields as a native interface. We often get requests from customers to show them some values related to their drop-down choice. E.g. they select a client record, displayed as "`<customerNo> <customerName>`", and want to see their contact details right away in some neighboring fields.

The computed field could be set up referencing the form field, eg. by using an `item`-reference, and the usual dot-notation, similar to @Voldemorten's suggestion:

Dropdown field: 
eg. `clientId` --> stores the client's primary id and displays "`<customerNo> <customerName>`"

Computed field `clientEmail` with two selectable types:
1. **Template/moustache syntax**: 
eg. `"Work email: {{ $item.clientId.email }}"`, of course with nested field lookups
2. **Javascript** (to allow for some basic calculations): 
eg. `"return 'Work email: ' + item.clientId.email;"`

Whenever `clientId` changes on the item, `clientEmail` should be updated accordingly. The field can be configured display-only or with its value actually stored in the db.

**Reply by amo42** [reply]
Shouldn't this be possible by using Flows and the Script operation? Still not that native as an actual interface for dedicated fields, but much simpler and easy to set up.

**Reply by stx-chris** [reply]
Thanks for your suggestion @amo42! Let me ponder on it 🤔

**Reply by stx-chris** [reply]
I managed to write a filter Event Hook trigger, followed by an Update Data operation, selected the currently active document and updated the computed field value within the payload, so I guess this workaround is the next-best-thing after a dedicated Computed Field interface 😉. 

<img width="860" alt="image" src="https://user-images.githubusercontent.com/62717406/187864047-68e64bfc-a2be-47c5-aa3f-c13130bf8108.png">

<img width="827" alt="image" src="https://user-images.githubusercontent.com/62717406/187864511-15dedd58-4729-4313-84d3-41c359931268.png">

**Reply by rijkvanzanten** [reply]
It sounds like your ideal setup for a computed field would be one that updates client side as well, isn't it? So it would update live on the page when you modify the dropdown? 🤔 

So far, I've only really thought about computed fields as "generated columns" in SQL, but that wouldn't live-update on the app side of things

**Comment by thegitcat** [comment]
I use this extension, does the job really well: https://github.com/rezo-labs/directus-extension-computed-interface

**Reply by alcompstudio** [reply]
Hello! I'm not a programmer! Can you please tell me if I can install this extension on the cloud version (like a https://my_project.directus.app/) ?

**Reply by br41nslug** [reply]
@alcompstudio You currently cannot yet install extensions on the Cloud. We're working on this but doing securely takes time.

**Reply by alcompstudio** [reply]
Thanks for your reply!

**Reply by faisal-abdin90** [reply]
This extension works perfectly if we are creating items in a collections form view but if I bulk upload a csv the calculated field wont get run the calculation.

In my use case for example I want to bulk upload data in a collection via csv import where i have date_of_birth and an empty string field where I would like to calculate the age from todays date. This extension will do it if I enter each record one by one on form view but wont calculate the age if I bulk upload the CSV.

I am not sure how to do this with flows (not very good with coding XD)

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by Jeffrey-H** [reply]
Whats the status? It has over 100 upvotes. Would love to see this feature :heart_eyes:

**Reply by rijkvanzanten** [reply]
TBD! It's a tricky one.. It's something that's supported on a large amount of databases, but not all. It's also not something that's available on other data stores we're aiming to support in the future, so the big question now is how to figure this out in way we can make it work natively on databases that support it, and fall back to a Directus implementation for the rest 🤔

**Reply by Jeffrey-H** [reply]
As the other plugin does above is that it gets calculated before save. On the Frontend. It's for sure better to calculate it on the database side for other applications. But indeed not all different data storages support these kinds of actions to be done in the database. Why not start with some data storages that support them and roll out new ones after a decision had been made of how to do it for the rest of those storages (or never support them at all). A solution thus can be to update on database save.

**Comment by Voldemorten** [comment]
Since this is getting some tracking once more, I just wanted to add my take on it.

I have created something, that somewhat achieves this, albeit it's a bit hacky. 

The idea is, that we have an interface to create the custom calculation fields and then we have a custom global hook that calculates these calculations values every time an item is updated and saves these to the database. It does this through custom javascript, that is written in the interface.

I think it's best explained through screenshots, which you can also find in this [repo](https://github.com/Voldemorten/directus-calculation-fields-extension), where you also have the complete code for the idea

This is an example with a third party api data where we retrieve some exchange rate from a third party api in our shared data. We use the name usd_to_dkk_mock
![usd_to_dkk_mock](https://github.com/directus/directus/assets/8776117/7f4a03e6-a934-46e7-959c-d9ff121355cb)

After the creation we have this in our shared data:
![shared_data_usd_to_dkk_mock](https://github.com/directus/directus/assets/8776117/41664785-8345-45a4-8968-434376eb496b)

This data is then accessible in another collection, that uses the shared item:
![calculation_field_in_action](https://github.com/directus/directus/assets/8776117/e7561174-ae6c-40cb-969c-878138014c99)
On the right side in the console you see that we have the item itself available in $item and the shared item in $shared. The shared item holds our value `usd_to_dkk_mock` which we can utilize in our new calculation field in another collection.

After we're done both fields will automatically update every 24 hours (a self-defined setting with a custom [cronjob](https://github.com/Voldemorten/directus-calculation-fields-extension/blob/main/global_item_hooks/src/index.js)) or whenever you update on of the items (this doesn't mean that both get updated. If you update shared_date the calculation fields in shared_data will be updated but not automatically the calculation fields in the other dependant collections and vice versa)

The interface supports a boolean called `automaticCalculation`. In the attached implementation of the hook, this should always be true. The reason that I created it, was because I wanted the users' to have an option to manually overwrite the result of the calculation, such that if the value was empty, the value of the calculation would be returned, and if not it would return the actual value. This required a read hook as well, which is not included in this repo. However, this turned out not be needed for us, so I removed support for it.

Let me know, if you need to be able to do this, and I'll look into providing it.

**BE CAREFUL WHEN USING THIS PLUGIN. BECAUSE IT USES CUSTOM JS CODE YOU CAN DO A LOT OF DAMAGE. THIS IS ONLY ATTENDED TO BE ACCESSIBLE TO ADMINS THAT KNOWS WHAT THEY ARE DOING**

**Reply by emahuni** [reply]
"Let me know, if you need to be able to do this, and I'll look into providing it."

yes please provide it. thanks

**Comment by emahuni** [comment]
I have this implementation that executes scripts per form that can be used to achieve this see discussion #19053 "Add scripting support for advanced users to supply scripts that allow them to control forms and fields with access to stores, apis etc. These scripts can be served by the Directus api using a /script endpoint. This allows app extensibility to a whole new level."

**Comment by w0kyj** [comment]
Related to generated database columns
- not all databases(data sources) support this
- data source capabilities need to allow or restrict functionality in app/api
    - flag fields read-only in Directus
    - configuration would be done on DB directly

API - javascript/code controlled
- not dependent on the database(data source)
- more options available
- handled by Directus (means not available in the data source)
    - formula structure and variables
- performance has to be considered and architected into the solution

**Comment by northern-mindset** [comment]
Is it possible to add few calculated fields to an API response using the [Custom API Endpoint](https://docs.directus.io/extensions/endpoints.html) extension type and the [`ItemsService`](https://docs.directus.io/extensions/services/accessing-items.html)?

Basically, I need all the fields from a collection plus a couple of calculated fields. Only through API.

**Reply by Voldemorten** [reply]
Yes. That it possible. However, if you need to use the ItemsService, I'm guessing that the calculated fields are directly related to the items, so maybe it would make better sense to create a custom hook (instead of an endpoint) where these fields are calculated.

**Reply by northern-mindset** [reply]
A `read` hook, right?

**Reply by Voldemorten** [reply]
Yes. That would be my way of solving it. But be aware that a read hook is executed every time you read an item, which can severely impact performance, so be a bit carefull about what you implement here as your calculation fields.

**Reply by northern-mindset** [reply]
The reason for the custom endpoint is that the calculated fields are needed only for a particular “export” for a search index. These calculated fields are kind of additional attributes for better search experience. For example, besides an ISO date stored in the collection, I want to add a human-formatted date (4 December 2003) to find “December”.

That is why the idea is to have a separate endpoint specifically for the “search export”. And no performance penalty.

Anyhow, thanks for your help!

**Reply by Voldemorten** [reply]
In that case I agree with your approach.

**Comment by jeffersonsolorzano** [comment]
Is there any update on this feature?

**Comment by alexheidl** [comment]
hi- am also interested if any update on this feature? reason is, we have a use case where i would like to calculate a checksum across a column inside a collection and present it to the content editor as they add or modify records

**Comment by phazonoverload** [comment]
Hello everyone! We released a calculated fields interface this week as an extension via the Directus Marketplace. 

It supports the full set of functions provided by [Formula.js](https://formulajs.info/) and a majority of JavaScript operators that work for numbers and strings. 

We're making sure to parse all formulas to make sure they only run allowed functions, and it supports M2O related fields. Just a note that values are only visible in the interface and not returned via the API.

We've released it as a bundle in anticipation that we may later add new extension types. Search for **Calculated Fields Bundle** in the Marketplace to find it in any Directus project, whether self-hosted or on Directus Cloud 😄 

![image](https://github.com/user-attachments/assets/5aa210f6-6a82-436d-9ab0-0731418b48ff)

**Reply by arladmin** [reply]
> Hello everyone! We released a calculated fields interface this week as an extension via the Directus Marketplace.
> ...

@phazonoverload 

Nice update!

Is there somewhere where the other part of this request is being already tracked (as this FR seems to have been closed)?
The other part being: persisting the calculated values in the DB and being able to use those values in API responses as well.

**Reply by phazonoverload** [reply]
There were a lot of discussions but the very initial request opened was just about an interface. We always welcome input, but will never be able to implement everything. By the time you're using the API, you should be in an environment where you can calculate values yourself, hence it wasn't implemented. 

In answer to your specific question though - no, it's not being tracked. If you want to see improvements to the extension, we welcome an issue on [directus-labs/extensions](https://github.com/directus-labs/extensions).

**Comment by arladmin** [comment]
> By the time you're using the API, you should be in an environment where you can calculate values yourself

I think that's quite different from this FR, at least from my perspective (and a rather insincere remark, if i may say so). 
Just because one is using the API doesn't imply that one is operating in a 'code' environment as well. Hence the need for being able to set 'calculation formulas' in the (no-code/low-code) Studio app.

My initial request, as well as that of quite a few others, was for persisting calculated values in the db (just like in Airtable etc.)

Nevertheless, for all those who'd like to support the full implementation of this feature, the FR is continued here: https://github.com/directus-labs/extensions/issues/73

**Reply by rijkvanzanten** [reply]
> I think that's quite different from this FR

And that's exactly why I started enforcing new feature requests to adhere to the RFC format. The original post here says absolutely nothing and leaves everything up to assumptions and interpretation 😞  The stricter RFC format prevents that by requiring feature requests to have a clearly defined definition of done and motivation of what goals we're trying to achieve. Way nicer to make sure everybody is aligned on what the ask is 🙂
